README.md
Palea
=====

Palea is a tool to help discover if devices on your (secured and firewalled) 
network are also unknowingly connected to unknown other networks that would
facilitate attacks and information leaks to the outside. Such an unknown 
network could for instance be a known device on your trusted network that 
also has a USB dongle in it connected to the open internet over GSM/2G/3G/xG.

By spoofing packets, Palea can be used to trick systems into exposing their
connections to the internet. Palea can be run 24/7 on your network to also
discover temporary connections.

##How-to

I assume you've read the PDF's on the techniques and goal of Palea. They
are also included in this git repository. These were written for a PoC I had 
before coding it all together.

Clone the repo and run ./thrower.py -h as root. You will get the help 
listing there. (make sure you have all the stuff you need, including Impacket)

In order to start throwing packages in the right directions with the right
options you'll need a file containing all the ip's you want to target. I've
included an nmap script in order to generate such a list from other notations.
Please open the list in your favourite text editor and ensure that your own
network and broadcast address are not in that list. Impacket, our packetlib,
appears to have issues with those. 

##How will you get the results from Palea? 

You've downloaded the source code together with a simple webinterface. Once
you've placed this on a server with apache or any other running webserver
you can be your own 'catcher'. You need to place this catcher in the net-
work (or internet) you want to detect leakages in. This solution however
will require some technical knowledge. In this paragraph I explain the
steps I took to get it al running on a brand new Debian installataion.
This example installation however doesn't deal with SSL nor with virtual
hosts. It assumes a stand alone Debian for this purpose only.

##Installation steps

    apt-get install openssh-server

    apt-get install screen python sqlite sqlite3 python-impacket python-pcapy 
    apache2 nmap openssl libssl0.9.8 libapache2-mod-php5 rdate php5-sqlite php-db 
    python-pysqlite2

    mkdir /home/www-data
    chown -R www-data:www-data /home/www-data
    unpack palea.tgz in /home/www-data it contains a dir www, set docroot to here
    vim /etc/apache2/sites-enabled/000-default
    edit line 4 and +/- 10.. change /var/www in /home/www-data/www
    apache2ctl restart
    chown -R www-data:www-data /home/www-data

    screen -mS catcher

*[Inside screen]*

    cd /home/www-data/palea-express
    ./catcher                                                     

[Now press CTRL+a then d to detach]

The following line gives you your IP's

    ifconfig|grep 'inet addr'|awk '{print $2}'|sed 's/addr://g'   #Get IP address

Use this IP in the launch of a heartbeat message (sesion 0)

    /home/www-data/palea-express/thrower.py -t I -s 0 -i /home/www-data/palea-express/iplists/hb.ips -c $ip

Now visit: http://$ip/palea/

You should see an overview with a 'heartbeat alive' message.


palea-express/CHANGELOG
-- VERSION 2.2 --

* cleaning:	Cleaning everything for first github release
* namechange: 	The software is renames to Palea as it approaches release
* bugfix:       option -o is not by default set to a file but to /dev/null, you 
                should not need this information. If you do please specify it 
                manually to a nonexisting file.
* readability:	Functions are now documented better. Also some code has 
                switched place, first functionality, thereafter verbose printing
                if needed.
                Help functions are now proper.

-- VERSION 2.1 --

Working release, no proper documentation

-- GENERAL INFORMATION --

Unpack and point your webserver to the www dir. This contains the reporting php scripts.
You will need apache, php, sqlite3, impacket for python and screen as packages.
You can have the server itself send out a ping to a random ip and have the catcher 
report this 'heartbeat'. Session 0 is used for this.

I'd advice to start the catcher.py as root in a screen...

We use nmap to generate an input list of IP's. in the iplists folder you'll find 
a script for this purpose. The thrower has a reasonable help function that should 
get you going. More documentation is to follow.

Put heartbeat in crontab as root:

# m h  dom mon dow   command
  * *  *   *   *    /home/www-data/virtual_host_1/palea-express/thrower.py -t I -s 0 -i /home/www-data/virtual_host_1/palea-express/iplists/hartbeat.ips -c "YOUR IP"


www/palea/index.php
<?php

$session = $_REQUEST['se'];

if ( $db = new PDO("sqlite:../../var/palea.db")){

	printf("<html><head>");
	printf("  <link href=\"style.css\" rel=\"stylesheet\" type=\"text/css\">");
	printf("  <title>Choose your session see results</title>");
	printf("  <meta http-equiv=\"refresh\" content=\"5\"/>");
	printf("</head>");
	printf("<body><table width=100%%><tr><td class=\"invisible\"><h1>Choose your session to see results</h1></td>");
	printf("<td class=\"invisible\" align=\"right\"><img src=\"logo.jpg\"></td></tr></table>");


// End of header, the value tables start here
	//$resultset = $db->query("SELECT timestamp, session, count(*) as count FROM catch GROUP BY session ORDER BY timestamp DESC;");
	$resultset = $db->query("SELECT timestamp, session, count(*) as count FROM catch WHERE gateway NOT LIKE \"127.0.0.1\" AND gateway NOT LIKE \"77.235.37.117\" GROUP BY session ORDER BY timestamp DESC;");
	printf("<table><tr><td class=\"invisible\" colspan=\"3\"><h2>Sessions and # of results</h2></td></tr>");
	printf("<tr><td class=\"head\">Last received packet on:</td><td class=\"head\">Remarks</td><td class=\"head\">Session</td>");
	printf("<td class=\"head\"># received packages</td><td class=\"head\">ACTION</td></tr>");
	$cnt = 1;
	while($result = $resultset->fetch()){
		$remark = "&nbsp;";
		if( ($result['session']  == 0) && ($result['timestamp'] > (time() -120) ) ){
			$remark = "<b><font color=\"00DD00\">Hartbeat! system alive!</font></b>";
		}elseif(  ($result['session']  == 0) ){
			$delay = time()-$result['timestamp'];
			settype($delay,"int");
			$remark = "<b><font color=\"FF0000\">ALARM no hartbeat recieved for " . $delay .  " seconds!</font>";
		}elseif( $result['timestamp'] > (time()-60) ){
			$remark = "<b><font color=\"FF0000\">Scan still running!</font></b>";
		}elseif( $result['timestamp'] < (time()-43200) ){
			 $remark = "<b><font color=\"CCCCCC\">Scan is older than 5 days!</font></b>";
		}

		printf("<tr><td> %s </td>",  strftime("%Y-%m-%d %H:%M:%S" ,$result['timestamp']) );
		printf("<td> %s </td>", $remark );
		printf("<td><a href=\"report.php?se=%s\">%s</a></td>", $result['session'], $result['session'] );
		printf("<td> %s </td>", $result['count'] );
		printf("<td>");
		printf("<a href=\"delete.php?se=%s\"><img src=delete.png border=0 width=16 alt=\"Delete\"></a>&nbsp", $result['session'] );
		printf("<a href=\"report.php?se=%s\"><img src=report.png border=0 width=16 alt=\"Report\"></a>&nbsp", $result['session'] );
		printf("</td></tr>");
	}
	printf("</table></body></html>");

}
?>


www/palea/report.php
<?php

$session = $_REQUEST['se'];

if ( $db = new PDO("sqlite:../../var/palea.db")){


	printf("<html><head><link href=\"style.css\" rel=\"stylesheet\" type=\"text/css\"><title>Results for session %d</title></head>", $session);
	printf("<body><table width=100%%><tr><td class=\"invisible\"><a href=\"index.php\">[home]</a><br><h1>Results for sesion %s</h1></td>", $session);
	printf("<td class=\"invisible\" align=\"right\"><img src=\"logo.jpg\"></td></tr></table>");

// End header section, start of tabular data)

	printf("<table><tr valign=\"TOP\"><td>");
	
	$resultset = $db->query("SELECT * FROM catch WHERE session = $session");
	printf("<table><tr><td class=\"invisible\" colspan=3><h2>All results </h2></td></tr>", $session);
	printf("<tr><td class=\"head\">Row id</td><td class=\"head\">Network station</td><td class=\"head\">Discovered gateway</td><td class=\"head\">Type</td></tr>");
	while($result = $resultset->fetch()){
		printf("<tr><td> %s </td><td>  %s </td><td> %s </td><td> %s </td></tr>", $result['uniqueId'], $result['victim'], $result['gateway'], $result['type'] );
	}
	printf("</table>");
	
	printf("</td><td>");

	$resultset = $db->query("SELECT gateway, count(*) as count FROM catch WHERE session = $session GROUP BY gateway ORDER BY count DESC;");
	printf("<table><tr><td class=\"invisible\" colspan=3><h2>Most found gateways</h2></td></tr>");
	printf("<tr><td class=\"head\">Result #</td><td class=\"head\">Gateway</td><td class=\"head\"># occurences</td></tr>");
	$cnt = 1;
	while($result = $resultset->fetch()){
		printf("<tr><td> %d </td><td> %s </td><td> %s </td></tr>", $cnt++ , $result['gateway'], $result['count']);
	}
	printf("</table></body></html>");

	printf("</td</tr></table>");
}
?>


palea-express/Logger.py
#!/usr/bin/env python

# See LICENSE.txt for details.
# Version see CHANGELOG file

import time
from pysqlite2 import dbapi2 as sqlite

class DB:
	""" Simple thin DB class for inserting into db """
	def __init__(self,filename='../var/palea.db'): #adjust this location of DB when needed
		self.connection = sqlite.connect(filename)
		self.cursor = self.connection.cursor()
		# We need to create a table for incoming packets
		self.cursor.execute('CREATE TABLE IF NOT EXISTS catch(id INTEGER PRIMARY KEY, timestamp REAL, victim VARCHAR(50), gateway VARCHAR(50), session INT, uniqueId REAL, type VARCHAR(50) )' )	
	
	def insert(self,table,victim,source,session,uniqueId,type):
                """ Insert a packet into the database """
		ts = time.time()
		sql = "INSERT INTO %s VALUES ( NULL,%f,?,?,?,?,? )" % (table, ts)
		self.cursor.execute(sql,( victim,source,session,uniqueId,type ) )
		return ts

	def close(self):
                """ commit before closing """
		self.connection.commit()


palea-express/catcher.py
#!/usr/bin/env python

# See LICENSE.txt for details.
# Version: see CHANGELOG file

# Importing default stuff
import select
import socket
import time
import sys
import struct
import getopt
import signal
import types
import string

# Importing specific items. Impacket should be on the system
from impacket import ImpactDecoder, ImpactPacket
from Logger import DB

# Global flag for loop, has to be turned on
RUNNING = 0

def usage():
        """ help function """
        print """
[version 2.2]
The catcher listenes for ICMP reply, Destination Unreachable and Port Unreach
-able packages. it stores some information from these packages in a sqlite DB.
This database can be queried with the simple php frontend.

! The catcher needs to be run with root privileges !

We advise to run the catcher in a screen (http://www.gnu.org/software/screen/)
so you can easily see if it's still running and whether packages are received.
Example:


Optional settings:
-h  --help          Shows this help message
-v                  Adds one to verbosity, use multiple times for more info
-q  --quiet         Default verbosity is 1, this sets it to 0, you will get no
                    output. This option cannot be used in combination with -v.
-D  --Debug         Activates debugging, all packets recieved will be logged to
                    a debug directory."""

def HandleSignal(sig, frame):
        """ Signal handling for quitting with CTRL - C"""
	global RUNNING
	if sig in (signal.SIGINT,signal.SIGTERM):
		#tell main loop to quit
		RUNNING = 0
	else:
		#ignore other signals
		print "SIGNAL PASSED"
		pass

def CatchICMP(conf):
        """ Main loop for catching packets """
	global RUNNING

	# Open a raw socket. Special permissions are usually required.
	s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_ICMP)
	s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)

	# Open a database object, the object will initialize a full db structure.
	datalog = DB()

	# Timestamp to start with
	ts_main = time.time()

	# Set some signals.
	oldInt = signal.signal(signal.SIGINT,HandleSignal)
	oldTerm = signal.signal(signal.SIGTERM,HandleSignal)
	oldHup = signal.signal(signal.SIGHUP,HandleSignal)

        # We are entering the main indefinite loop
	RUNNING = 1
	while RUNNING:
		# Wait for incoming replies.
		if s in select.select([s],[],[],1)[0]:
			reply = s.recvfrom(2000)[0]

			# Use ImpactDecoder to reconstruct the packet hierarchy.
			rip = ImpactDecoder.IPDecoder().decode(reply)
			# Extract the ICMP packet from its container (the IP packet).
			ricmp = rip.child()
			
			################################
			# Types of received packages
			# 1, ICMP Reply
			# 2, DU ICMP
			# 3, DU UDP
			# TODO: DU PU UDP (Port unreach)
			################################
	
			# If type is ICMP Reply
			if ricmp.ICMP_ECHOREPLY == ricmp.get_icmp_type():
				data = ricmp.get_data_as_string()
				insert_ts = datalog.insert( 'catch', socket.inet_ntoa(data[0:4]), rip.get_ip_src(),\
						socket.ntohs(struct.unpack('H',data[4:6])[0]), socket.ntohs(struct.unpack('H',data[6:8])[0]), "ICMP REPLY" )
                                
                                # block of code for writing verbose output
				if conf['verbose'] in range(1,3):
					sys.stdout.write(".")
					sys.stdout.flush()
				if conf['verbose'] > 2:
					print "Session:%6s | Victim:%16s | Gateway:%16s | Unique:%16f | Type:%12s" % \
						(socket.ntohs(struct.unpack('H',data[4:6])[0]),socket.inet_ntoa(data[0:4]),rip.get_ip_src(),\
							socket.ntohs(struct.unpack('H',data[6:8])[0]), "ICMP REPLY")

			# If we're dealing with a DU packet		
			if ricmp.ICMP_UNREACH == ricmp.get_icmp_type():
				data = ricmp.get_data_as_string()

                                # If the DU is a reply to ICMP	
				if  struct.unpack('c',data[9:10])[0] == '\x01':
					insert_ts = datalog.insert( 'catch', socket.inet_ntoa(data[16:20]), rip.get_ip_src(),\
						socket.ntohs(struct.unpack('H',data[32:34])[0]), socket.ntohs(struct.unpack('H',data[34:36])[0]), "DU ICMP" )
                                        
                                        # block of code for writing verbose output
					if conf['verbose'] in range(1,3):
						sys.stdout.write(".")
						sys.stdout.flush()
                                	if conf['verbose'] > 2:
						print "Session:%6s | Victim:%16s | Gateway:%16s | Unique:%16f | Type:%12s" % \
							(socket.ntohs(struct.unpack('H',data[32:34])[0]),socket.inet_ntoa(data[16:20]),rip.get_ip_src(),\
								socket.ntohs(struct.unpack('H',data[34:36])[0]), "DU ICMP" )

                                # If the DU is a reply to UDP
				elif struct.unpack('c',data[9:10])[0] == '\x11':
					insert_ts = datalog.insert( 'catch', socket.inet_ntoa(data[16:20]), rip.get_ip_src(),\
				        	socket.ntohs(struct.unpack('H',data[20:22])[0]), socket.ntohs(struct.unpack('H',data[4:6])[0]), "DU UDP" )

                                        # block of code for writing verbose output
                                	if conf['verbose'] in range(1,3):
						sys.stdout.write(".")
						sys.stdout.flush()
                                	if conf['verbose'] > 2:
						print "Session:%6s | Victim:%16s | Gateway:%16s | Unique:%16f | Type:%12s" % \
							(socket.ntohs(struct.unpack('H',data[20:22])[0]),socket.inet_ntoa(data[16:20]),rip.get_ip_src(),\
								socket.ntohs(struct.unpack('H',data[4:6])[0]), "DU UDP" )

				else:
					print struct.unpack('c',data[9:10])
					print ("consider implementing this new type of DU package sinds we hit it.")

                        # If we're in debug mode we need to do more	
			if conf['debug']:
				# We need to open a new file and write the full packet to it
				fp = open("debug/%f.pckt" % insert_ts , "w")
				fp.write( reply )
				fp.flush()
				fp.close()

		# We only commit to database when needed for performance reasons
		if( (time.time()-ts_main) > 10):
			datalog.connection.commit()
			ts_main = time.time()
			
                        # block of code for writing verbose output
                        if conf['verbose'] in range (2,3):
                        	sys.stdout.write("#")
                        	sys.stdout.flush()
			if conf['verbose'] > 2:
				print "  database commit done, 10 seconds since last commit have passed!"

	print "Succesfully stopped"
	# Reset the signalhandlers
	signal.signal(signal.SIGINT,oldInt)
	signal.signal(signal.SIGINT,oldTerm)
	signal.signal(signal.SIGINT,oldHup)
	
	datalog.close()

def main(argv):
        """Main function, reads arguments, does sanity checks and will start server loop """

	# Some default configuration value's
	conf = {"verbose":1,"debug":0,"quiet":0}

        # Code block for handeling options
        try:
                optlist, list = getopt.getopt(argv[1:],"qhvD", ["quiet","help","debug"])
        except getopt.GetoptError, err:
                # Print helpinfo and exit
                print str(err) # Will tell what option isn't recognized
                usage()
                sys.exit(1)
        for opt in optlist:
                if opt[0] == "-v":
                        conf["verbose"] = conf["verbose"] + 1
                elif opt[0] in ("-q","--quiet"):
                        conf["quiet"] = True
                elif opt[0] in ("-D","--debug"):
                        conf["debug"] = 1
                elif opt[0] in ("-h","--help"):
                        usage()
                        sys.exit(0)

	# Sanity check, we only have one! Default verbose is 1, quiet makes it 0. -v and -q should not be used together
	if conf['quiet']:
		if conf['verbose'] > 1:
			print "Please don't use -q or --quiet in combination with -v"
			sys.exit(0)
		conf['verbose'] = 0

        # Block of code for writing verbose output
	if conf['verbose'] > 0: 
		print "Catcher running with verbosity %d and debug-level %d" % (conf['verbose'],conf['debug'])

	# Start the server loop
	CatchICMP(conf)

if __name__ == "__main__":
        main(sys.argv)



palea-express/thrower.py
#!/usr/bin/env python

# See LICENSE.txt for details.
# Version See CHANGELOG file

# imports of default stuff
import os
import signal
import string
import urllib
import types
import getopt
import struct
import select
import socket
import time
import sys

# importing impacket
from impacket import ImpactDecoder, ImpactPacket

# importing own function
import Functions as f

def usage():
        """ help function, might need an update """
        print """
[version 2.2]
The thrower sends out packages to systems that will try and reply to the 
palea catcher. 
Examepl:
thrower.py -vv -d 0.3 -t UI -s 1 -i ./iplists/hartbeat.ips -c "YOUR IP"
Atleast use the options -i -s and -c

Required settings:
-i  --input=        this should point to the list of ip's you wan't to target
-s  --session=      The sessionnumber (to identify your packages when recieved
                    by the cathcer. choose between 1 and 65 535, 0 may be used 
                    as a hartbeat monitor.

Optional settings:
-c  --catcher=      this is the ipadress of the catcher.
-d  --delay=        A delay may be specified for slow networks, you don't wan't
                    to overload the network, preventing packages from reaching
                    it's target.
-h  --help=         Shows this help message
-o  --output=       this points to the output file for a logging of all packets
                    sent during this session
-t  --tests=        U for UDP, I for ICMP or UI for both.
-v                  Adds one to verbosity, use multiple times for more info
"""

def SendUDP(srcIP,dstIP,sessionNr,counter):
        """ SendUDP sends a specially crafted UDP packet """

        # prepare the IP part
	ip = ImpactPacket.IP()
	ip.set_ip_src(srcIP)
	ip.set_ip_dst(dstIP)
	ip.set_ip_id(counter)
	
        # prepare the ICMP part
	udp = ImpactPacket.UDP()
	udp.set_uh_sport(sessionNr)
	udp.set_uh_dport(53)

	#auto generate checksum
	udp.set_uh_sum(0)
	udp.auto_checksum = 1

        # prepare the payload
        # put the target IP and the session number in the payload also for later recovery
        data = socket.inet_aton(dstIP)+struct.pack('H',socket.htons(sessionNr))+struct.pack('H',socket.htons(counter))
	
        # compose the total packet IP / icmp / payload
        udp.contains(ImpactPacket.Data(data))
	ip.contains(udp)

	# Open a raw socket. Special permissions are usually required.
	s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_UDP)
	s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)

	# and set it free
	s.sendto(ip.get_packet(), (dstIP, 0))
	
	# return timestamp for further reference
	return time.time()

def SendICMP(srcIP,dstIP,sessionNr,counter):
        """ SendICMP sends a specially crafted ICMP packet """
        
        # prepare the IP part
	ip = ImpactPacket.IP()
	ip.set_ip_src(srcIP)
	ip.set_ip_dst(dstIP)
	#this counter isn't used.
	ip.set_ip_id(counter)
	
        # prepare the ICMP part
	icmp = ImpactPacket.ICMP()
	#is used to read out uniquenumber in case of DU ICMP reply
	icmp.set_icmp_id(sessionNr)
	#is used to read out sessionnumber in case of DU ICMP reply
	icmp.set_icmp_seq(counter)

	#auto generate checksum
	icmp.set_icmp_cksum(0)
	icmp.auto_checksum = 1
	icmp.set_icmp_type(icmp.ICMP_ECHO)

        # prepare the payload
        # put the target IP and the sequence number in the payload also for later recovery
        data = socket.inet_aton(dstIP)+struct.pack('H',socket.htons(sessionNr))+struct.pack('H',socket.htons(counter))
	
	# compose the total packet IP / icmp / payload
        icmp.contains(ImpactPacket.Data(data))
	ip.contains(icmp)

	# Open a raw socket. Special permissions are usually required.
	s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_ICMP)
	s.setsockopt(socket.IPPROTO_IP, socket.IP_HDRINCL, 1)

	# and set it free
	s.sendto(ip.get_packet(), (dstIP, 0))
	
	# return timestamp for further reference
	return time.time()

def main(argv):
        """ main function shooting out packets based on a list of ip's and commandline options """

	#let's keep some config values in conf dict
	conf = {"verbose":0,"session":"NULL","inputfile":"iplist.txt","outputfile":"NULL","catcher":"NULL","tests":"IU","delay":0}
        "FaF: let's shoot the buggers out"

	# Lets catch the starttime early
	starttime = time.time()

        # Gathering all the commandline options and process them in the conf dict.
	try:
		optlist, list = getopt.getopt(argv[1:],"hvs:i:c:o:t:d:", ["help","session=","input=","catcher=","output=","tests=","delay="])
	except getopt.GetoptError, err:
		#print helpinfo and exit
		print str(err) #will tell what option isn't recognized
		usage()
		sys.exit(1)
	for opt in optlist:
		if opt[0] == "-v":
			conf["verbose"] = conf["verbose"] + 1
		elif opt[0] in ("-s","--session"):
			conf["session"] = opt[1]
		elif opt[0] in ("-i","--input"):
			conf["inputfile"] = opt[1]
		elif opt[0] in ("-o","--output"):
			conf["outputfile"] = opt[1]
		elif opt[0] in ("-c","--catcher"):
			conf["catcher"] = opt[1]
		elif opt[0] in ("-t","--tests"):
			conf["tests"] = opt[1]
		elif opt[0] in ("-d","--delay"):
			conf["delay"] = opt[1]
		elif opt[0] in ("-h","--help"):
			usage()
			sys.exit(0)

        # Some information for our verbose users
	if conf["verbose"] > 1:
		print "----------------------------------------"
		print "-- Starting thrower"
		print ""
		print "Using parameters:"
		print "- verbosity  is: %s" % (conf["verbose"])
		print "- session    is: %s" % (conf["session"])
		print "- delay      is: %s" % (conf["delay"])
		print "- inputfile  is: %s" % (conf["inputfile"])
		print "- outputfile is: %s" % (conf["outputfile"])
		print "- catcher    is: %s" % (conf["catcher"])
		print "- tests     are: %s" % (conf["tests"])
		print ""	
		print "----------------------------------------"
	
	
	# We've gonna do some sanity checks here, lets use a bool to see if we need to quit
	# DANGER: we assume all is ok until we decide it's not, actually we want 'deny by default'
	quit = False
	
        # Recent hack to ensure we can work without output (for the heartbeat)
        if (conf['outputfile'] is "NULL"):
                conf['outputfile'] = "/dev/null"
	# test is output file exists and then add extention number if file exists
	elif os.path.exists(conf["outputfile"]):
		ext_nr = 1
		while os.path.exists(conf["outputfile"] + "." + str(ext_nr)):
			ext_nr = ext_nr +1
		conf["outputfile"] = conf["outputfile"] + "." + str(ext_nr)
		
	# Opening list of IP's
	try:		
		input_fp = open(conf["inputfile"], "r")
	except IOError, (ErrorNumber,ErrorMessage):
		if ErrorNumber == 2:    # file not found, most likely
			print "- Sorry, file \"%s\" not found, please feed me a file with only ipaddresses" % (conf["inputfile"])
		else:                   # some other file error triggered
			print "- Congrats, you managed to trigger errornumber %s" % (ErrorNumber)
			print "     " + ErrorMessage
		sys.exit(1)
	if conf["verbose"] > 1 : print "- %s opened" % conf["inputfile"]

	# Validate IP for catcher
	if f.isValidIp(conf["catcher"]) == False :
		print "- %s (Catcher) is not a valid IP address!" % (conf["catcher"])
		quit = True

        # Sanity checking on some numeric values
	try:
		float(conf["delay"])
		if (float(conf["delay"]) < 0 ) or (float(conf["delay"]) > 60 ) :
			raise Exception("Out of range! timing needs to be between 0.000 and 60.000 seconds.")
	except:
		print "- %s is not a valid delay value in seconds, choose between 0.000 and 60.000 seconds." % (conf["delay"])
		quit = True

	try:
		int(conf["session"])
		if int(conf["session"]) not in range(0,65535):
			raise Exception("Out of range!")
	except:
		print "- %s is not a valid session ID, choose between 0 and 65535." % (conf["session"])
		quit = True
		
	# Read the IP file and test all ip's for validity	
	ips = input_fp.readlines()
	input_fp.close()

	for ip in ips:
		#check if all ip's provided are ok
		if f.isValidIp(ip) == False :
			quit = True
			print "- %s is not a valid IP address!" % (ip.strip())
		
	# Opening the determined output filename if we're not quitting already
	if quit == False:
		try:		
			output_fp = open(conf["outputfile"], "w")
		except IOError, (ErrorNumber,ErrorMessage):
			print "- Congrats, you managed to trigger errornumber %s" % (ErrorNumber)
			print "     " + ErrorMessage
			sys.exit(1)
		if conf["verbose"] > 1 : print "- %s opened" % conf["outputfile"]

	# See if real tests are send out, else we need to tell user it is a dry run
	if ("U" not in conf["tests"]) and ("I" not in conf["tests"]):
		print "No tests selected! use -t IU  (ICMP UDP)"
		quit = True
	
	# Number of checks done, lets see if we need to quit	
	if quit:
		print ""
                print "One of more errors in your command line options"
		print "Please get your stuff together! I'm quitting!"
		sys.exit(1)

        # We set a counter to keep count of packets send
	counter = 0
 	for ip in ips:	
		time.sleep(float(conf["delay"]))
                # If we configured I, we need to send ICMP packets
		if "I" in conf["tests"]:
			counter = counter + 1
			unique_id = SendICMP(conf["catcher"], ip, int(conf["session"]), counter )
               		# writing csv    id;unique;session;catcher;victim;
                	output_fp.write("NULL;%f;%s;%s;%s;%s\n" % (unique_id,conf["session"],conf["catcher"],ip.strip(),"ICMP") )
			if conf["verbose"] > 2:
				print "  - timstamp: %10f | victim: %16s | catcher: %16s | session: %4s | unique: %16f | test:%5s"\
					% (unique_id,ip.strip(),conf["catcher"],conf["session"],counter,"ICMP")

                # IF we configured U, we need to send UDP packets
		if "U" in conf["tests"]:
			counter = counter +1
			unique_id = SendUDP(conf["catcher"], ip, int(conf["session"]), counter )
        	        # writing csv    id;unique;session;catcher;victim;
	                output_fp.write("NULL;%f;%s;%s;%s;%s\n" % (unique_id,conf["session"],conf["catcher"],ip.strip(),"UDP") )
        	        if conf["verbose"] > 2:
                	        print "  - timstamp: %10f | victim: %16s | catcher: %16s | session: %4s | unique: %16f | test:%5s" \
					% (unique_id,ip.strip(),conf["catcher"],conf["session"],counter,"UDP")

	        # Verbosety: printing a U or an I for each packet sent.	
		if conf["verbose"] in range(1,3):
			sys.stdout.write(conf["tests"])
			sys.stdout.flush()
		# If our counter approaches the max it will reset, NOTE received reply's might not be unique anymore!
                if counter > 65000: 
			counter = 1
                        # Verbosety printing
			if conf["verbose"] in range(1,3): sys.stdout.write("1")
			if conf["verbose"] > 2: print " - Resetting unique counter"
	if conf["verbose"] > 0 : print "\nFinished sending %d packets in %s seconds" % (counter , (time.time() - starttime))
		
if __name__ == "__main__":
        main(sys.argv)
